File: Lecture 1 - C/Problem Set 1/Credit/credit.c
/*
    PROBLEM: CREDIT CARD VALIDATOR (STANDARD C VERSION)

    Validate a credit card number using Luhn’s Algorithm
    and determine card type.

    No CS50 library used.
*/

#include <stdio.h>
#include <stdbool.h>

// Function prototypes
long get_number(void);
bool luhn_algo(long number);
int get_length(long number);
int get_start(long number, int digits);

int main(void)
{
    long number = get_number();

    if (!luhn_algo(number))
    {
        printf("INVALID\n");
        return 0;
    }

    int length = get_length(number);
    int start1 = get_start(number, 1);
    int start2 = get_start(number, 2);

    if ((length == 15) && (start2 == 34 || start2 == 37))
    {
        printf("AMEX\n");
    }
    else if ((length == 16) && (start2 >= 51 && start2 <= 55))
    {
        printf("MASTERCARD\n");
    }
    else if ((length == 13 || length == 16) && start1 == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}

long get_number(void)
{
    long n;

    do
    {
        printf("Number: ");
        scanf("%ld", &n);
    }
    while (n <= 0);

    return n;
}

bool luhn_algo(long number)
{
    int sum = 0;
    bool alternate = false;

    while (number > 0)
    {
        int digit = number % 10;

        if (alternate)
        {
            digit *= 2;

            if (digit > 9)
            {
                digit -= 9;
            }
        }

        sum += digit;
        alternate = !alternate;
        number /= 10;
    }

    return (sum % 10 == 0);
}

int get_length(long number)
{
    int count = 0;

    while (number > 0)
    {
        number /= 10;
        count++;
    }

    return count;
}

int get_start(long number, int digits)
{
    while (number >= 10 && get_length(number) > digits)
    {
        number /= 10;
    }

    return number;
}

/*
---------------- ALGORITHM ----------------

1. Prompt user for card number.
2. Validate using Luhn’s Algorithm:
    - Double every second digit from right.
    - Subtract 9 if result > 9.
    - Sum all digits.
    - Check if sum % 10 == 0.
3. If invalid → print INVALID.
4. Count total digits.
5. Extract starting digits.
6. Determine card type using rules.
7. Print card name.
8. End program.

Time Complexity: O(n)
Space Complexity: O(1)
*/

File: Lecture 6 - Python/Problem Set 6/Credit/credit.py
"""
PROBLEM: CREDIT CARD VALIDATOR

Write a program that:
1. Prompts the user for a positive credit card number.
2. Uses Luhn’s Algorithm to validate the number.
3. Determines the card type:
      - AMEX (15 digits, starts with 34 or 37)
      - MASTERCARD (16 digits, starts with 51–55)
      - VISA (13 or 16 digits, starts with 4)
4. Prints the card type or INVALID.
"""


# Get positive integer from user
def get_number():
    while True:
        try:
            number = int(input("Number: "))
            if number > 0:
                return number
            else:
                print("Give number greater than zero.")
        except ValueError:
            print("Give number greater than zero.")


# Apply Luhn's Algorithm
def luhn_algo(num):
    total = 0
    alternate = False
    temp_number = num

    while temp_number > 0:
        digit = temp_number % 10

        if alternate:
            digit *= 2
            if digit > 9:
                digit -= 9

        total += digit
        alternate = not alternate
        temp_number //= 10

    return total % 10 == 0


# Get length of number
def get_length(num):
    count = 0
    while num > 0:
        num //= 10
        count += 1
    return count


# Get first n digits
def get_start(num, digits):
    length = get_length(num)

    while length > digits:
        num //= 10
        length -= 1

    return num


# Main logic
number = get_number()

# Validate using Luhn
if not luhn_algo(number):
    print("INVALID")
else:
    length = get_length(number)

    start1 = get_start(number, 1)
    start2 = get_start(number, 2)

    if length == 15 and (start2 == 34 or start2 == 37):
        print("AMEX")
    elif length == 16 and 51 <= start2 <= 55:
        print("MASTERCARD")
    elif (length == 13 or length == 16) and start1 == 4:
        print("VISA")
    else:
        print("INVALID")


"""
---------------- ALGORITHM ----------------

1. Prompt user for positive integer.
2. Apply Luhn’s Algorithm:
      a. Starting from last digit, alternate digits.
      b. Double every second digit.
      c. If doubled digit > 9, subtract 9.
      d. Sum all digits.
      e. If total % 10 == 0 → valid.
3. If invalid → print INVALID.
4. If valid:
      a. Get total length.
      b. Get first digit and first two digits.
      c. Check card rules:
            - AMEX → 15 digits & starts with 34 or 37
            - MASTERCARD → 16 digits & starts with 51–55
            - VISA → 13 or 16 digits & starts with 4
5. Print card type or INVALID.

Time Complexity: O(n)
Space Complexity: O(1)
"""